File: sdk/dotnet/Mics.Client/AesGcmMessageCrypto.cs
using System.Security.Cryptography;

namespace Mics.Client;

public sealed class AesGcmMessageCrypto : IMicsMessageCrypto
{
    private const byte Version = 1;
    private const int NonceBytes = 12;
    private const int TagBytes = 16;

    private readonly byte[] _key;

    public AesGcmMessageCrypto(byte[] key)
    {
        ArgumentNullException.ThrowIfNull(key);
        if (key.Length is not (16 or 24 or 32))
        {
            throw new ArgumentException("AES key length must be 16/24/32 bytes.", nameof(key));
        }

        _key = key.ToArray();
    }

    public byte[] Encrypt(ReadOnlySpan<byte> plaintext)
    {
        if (plaintext.Length == 0)
        {
            return Array.Empty<byte>();
        }

        var nonce = new byte[NonceBytes];
        RandomNumberGenerator.Fill(nonce);

        var ciphertext = new byte[plaintext.Length];
        var tag = new byte[TagBytes];

        using var aes = new AesGcm(_key, TagBytes);
        aes.Encrypt(nonce, plaintext, ciphertext, tag);

        var output = new byte[1 + NonceBytes + TagBytes + ciphertext.Length];
        output[0] = Version;
        Buffer.BlockCopy(nonce, 0, output, 1, NonceBytes);
        Buffer.BlockCopy(tag, 0, output, 1 + NonceBytes, TagBytes);
        Buffer.BlockCopy(ciphertext, 0, output, 1 + NonceBytes + TagBytes, ciphertext.Length);
        return output;
    }

    public byte[] Decrypt(ReadOnlySpan<byte> ciphertext)
    {
        if (ciphertext.Length == 0)
        {
            return Array.Empty<byte>();
        }

        if (ciphertext.Length < 1 + NonceBytes + TagBytes)
        {
            throw new CryptographicException("Ciphertext too short.");
        }

        if (ciphertext[0] != Version)
        {
            throw new CryptographicException("Unsupported ciphertext version.");
        }

        var nonce = ciphertext.Slice(1, NonceBytes);
        var tag = ciphertext.Slice(1 + NonceBytes, TagBytes);
        var enc = ciphertext.Slice(1 + NonceBytes + TagBytes);

        var plaintext = new byte[enc.Length];
        using var aes = new AesGcm(_key, TagBytes);
        aes.Decrypt(nonce, enc, tag, plaintext);
        return plaintext;
    }
}

File: sdk/android/mics-client-sdk/mics-client-sdk/src/main/kotlin/com/mics/clientsdk/AesGcmMessageCrypto.kt
package com.mics.clientsdk

import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

class AesGcmMessageCrypto(rawKey: ByteArray) : MessageCrypto {
    init {
        require(rawKey.size == 16 || rawKey.size == 24 || rawKey.size == 32) {
            "AES key length must be 16/24/32 bytes"
        }
    }

    private val key = SecretKeySpec(rawKey, "AES")
    private val rnd = SecureRandom()

    override fun encrypt(plaintext: ByteArray): ByteArray {
        if (plaintext.isEmpty()) return ByteArray(0)

        val nonce = ByteArray(NONCE_BYTES)
        rnd.nextBytes(nonce)

        val cipher = Cipher.getInstance(TRANSFORM)
        cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(TAG_BITS, nonce))
        val sealed = cipher.doFinal(plaintext)

        val ciphertextLen = sealed.size - TAG_BYTES
        val ciphertext = sealed.copyOfRange(0, ciphertextLen)
        val tag = sealed.copyOfRange(ciphertextLen, sealed.size)

        val out = ByteArray(1 + NONCE_BYTES + TAG_BYTES + ciphertext.size)
        out[0] = VERSION
        nonce.copyInto(out, destinationOffset = 1)
        tag.copyInto(out, destinationOffset = 1 + NONCE_BYTES)
        ciphertext.copyInto(out, destinationOffset = 1 + NONCE_BYTES + TAG_BYTES)
        return out
    }

    override fun decrypt(ciphertext: ByteArray): ByteArray {
        if (ciphertext.isEmpty()) return ByteArray(0)
        if (ciphertext.size < 1 + NONCE_BYTES + TAG_BYTES) error("ciphertext too short")
        if (ciphertext[0] != VERSION) error("unsupported ciphertext version")

        val nonce = ciphertext.copyOfRange(1, 1 + NONCE_BYTES)
        val tag = ciphertext.copyOfRange(1 + NONCE_BYTES, 1 + NONCE_BYTES + TAG_BYTES)
        val enc = ciphertext.copyOfRange(1 + NONCE_BYTES + TAG_BYTES, ciphertext.size)

        val combined = ByteArray(enc.size + TAG_BYTES)
        enc.copyInto(combined, 0)
        tag.copyInto(combined, enc.size)

        val cipher = Cipher.getInstance(TRANSFORM)
        cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(TAG_BITS, nonce))
        return cipher.doFinal(combined)
    }

    companion object {
        private const val VERSION: Byte = 1
        private const val NONCE_BYTES = 12
        private const val TAG_BYTES = 16
        private const val TAG_BITS = TAG_BYTES * 8
        private const val TRANSFORM = "AES/GCM/NoPadding"
    }
}
